### PR TITLE
Improve message template construction performance

### PR DESCRIPTION
**Pull request by adamchester**
By using a new method, similar to `Enumerable.OfType<TResult>` but faster.

My tests show it's ~2X or ~3X faster.

### CODE REVIEW

**Comment by adamchester** [comment]
Benchmark is [here](https://gist.github.com/adamchester/57fa05c082b00354595532c0e6b1c79e).

**Comment by adamchester** [comment]
Note that since serilog caches message templates, this will not impact normal logging speed much.

File: src/Serilog/Events/MessageTemplate.cs [hunk 1]
```diff
@@ -84,6 +84,24 @@ public MessageTemplate(string text, IEnumerable<MessageTemplateToken> tokens)
         }
 
         /// <summary>
+        /// Similar to <see cref="Enumerable.OfType{TResult}"/>, but faster.
+        /// </summary>
+        private static TResult[] GetElementsOfTypeToArray<TResult>(object[] tokens)
```
Comment (nblumhardt) [inline src/Serilog/Events/MessageTemplate.cs]: Explicit `private` (just keen to keep the codebase consistent around this).

**Comment by nblumhardt** [comment]
Very nice! Guessing the initial size hint is the key to the perf boost :+1: 

While it won't make much difference to normal logging perf, in other scenarios (and in apps where the template cache is exhausted) it'll be worthwhile.

Just the one nitpick, otherwise LGTM.

**Comment by adamchester** [comment]
Actually, I think the key is that LINQ does many unnecessary things, like `IDisposable` + `Iterator` instances and so on. I figured `/2` was a good size default because it seems like most messages would have a property token combined with text tokens.

**Pull request merged**